pytripgui/add_vois_vc/add_vois_cont.py

```diff
@@ -0,0 +1,109 @@
+
+logger = logging.getLogger(__name__)
+
+
+class AddVOIsController:
+    def __init__(self, model, view):
+        self.model = model
+        self.view = view
+        self.is_accepted = False
+        self._setup_callbacks()
+
+        self._set_view_from_model()
+
+    def _setup_callbacks(self):
+        self.view.accept_buttons.accepted.disconnect()
+        self.view.accept_buttons.accepted.connect(self._save_and_exit)
+
+        self.view.add_voi_button.emit_on_click(lambda: self._create_add_voi_dialog())
+
+    def _save_and_exit(self):
+        self._set_model_from_view()
+        self.is_accepted = True
+        self.view.accept()
+
+    def _create_add_voi_dialog(self):
+        view = AddSingleVOIQtView()
+        controller = AddSingleVOIController(self.model.ctx, view)
+
+        view.show()
+
+        if not controller.is_accepted:
+            return
+
+        voi_widget = controller.get_voi_widget()
+        voi_widget.disable_fields()
+        list_element_voi = ListElementVOI(voi_widget)
+        self.view.voi_scroll_area.widget().layout().insertWidget(0, list_element_voi)
+
+    def _set_view_from_model(self):
+        ctx = self.model.ctx
+        view = self.view
+
+        view.name.text = ctx.patient_name
+        view.width.text = ctx.dimx * ctx.pixel_size
+        view.height.text = ctx.dimy * ctx.pixel_size
+        view.depth.text = ctx.slice_number * ctx.slice_distance
+        view.pixel_size.text = ctx.pixel_size
+        view.pixel_number_x.text = ctx.dimx
+        view.pixel_number_y.text = ctx.dimy
+        view.slice_number.text = ctx.slice_number
+        view.slice_distance.text = ctx.slice_distance
+        # TODO offsets
+        view.slice_offset.text = ctx.zoffset
+
+    def _set_model_from_view(self):
+        ctx = self.model.ctx
+        vdx = self.model.vdx
+
+        voi_widgets = self.view.voi_scroll_area.widget().layout()
+        for index in range(voi_widgets.count() - 1):
+            voi_widget = voi_widgets.itemAt(index).widget()
+
+            if isinstance(voi_widget, SphericalVOIWidget):
+                voi = create_sphere(
+                    cube=ctx,
+                    name=voi_widget.name,
+                    center=voi_widget.center,
+                    radius=voi_widget.radius,
+                )
+            elif isinstance(voi_widget, CuboidalVOIWidget):
+                voi = create_cube(
+                    cube=ctx,
+                    name=voi_widget.name,
+                    center=voi_widget.center,
+                    width=voi_widget.width,
+                    height=voi_widget.height,
+                    depth=voi_widget.depth,
+                )
+            else:
+                logger.debug("VOI widget unrecognised")
+                return
+            vdx.add_voi(voi)
+
+
+class ListElementVOI(QtWidgets.QFrame):
+    def __init__(self, voi_widget):
+        super().__init__()
+        path = os.path.join(os.path.curdir, "view", "list_element_voi.ui")
```

Comment:
Why `os.path.curdir` ?
Check solution used here https://github.com/pytrip/pytripgui/blob/52f7fc17263882887d505a705551bc45d330ef6d/pytripgui/view/qt_gui.py

Reply:
It is crashing on my mac:
```
(venv) Leszeks-Air:pytripgui grzanka$ python -m pytripgui.main
2021-09-16 14:27:10.826 python[36988:1789934] -[NSNull length]: unrecognized selector sent to instance 0x7fff80837eb0
2021-09-16 14:27:10.829 python[36988:1789934] Exception calling directoryDidChange: on delegate <QNSOpenSavePanelDelegate: 0x7fb71c368090> exception=-[NSNull length]: unrecognized selector sent to instance 0x7fff80837eb0
2021-09-16 14:27:10.830 python[36988:1789934] -[NSNull length]: unrecognized selector sent to instance 0x7fff80837eb0
2021-09-16 14:27:10.830 python[36988:1789934] Exception calling directoryDidChange: on delegate <QNSOpenSavePanelDelegate: 0x7fb71c368090> exception=-[NSNull length]: unrecognized selector sent to instance 0x7fff80837eb0
2021-09-16 14:27:13.452 python[36988:1789934] -[NSNull length]: unrecognized selector sent to instance 0x7fff80837eb0
2021-09-16 14:27:13.452 python[36988:1789934] Exception calling directoryDidChange: on delegate <QNSOpenSavePanelDelegate: 0x7fb71c368090> exception=-[NSNull length]: unrecognized selector sent to instance 0x7fff80837eb0
/Users/grzanka/PycharmProjects/pytripgui/pytripgui/canvas_vc/plotter/coordinate_info.py:32: MatplotlibDeprecationWarning: Axes3D(fig) adding itself to the figure is deprecated since 3.4. Pass the keyword argument auto_add_to_figure=False and use fig.add_axes(ax) to suppress this warning. The default value of auto_add_to_figure will change to False in mpl3.5 and True values will no longer work in 3.6.  This is consistent with other Axes classes.
  super().__init__(fig, rect, **kwargs)
/Users/grzanka/PycharmProjects/pytripgui/pytripgui/main.py:50: UserWarning: Tight layout not applied. tight_layout cannot make axes width small enough to accommodate all axes decorations
  return app.exec_()
QWidget::repaint: Recursive repaint detected
QWidget::paintEngine: Should no longer be called
QPainter::begin: Paint device returned engine == 0, type: 1
QPainter::end: Painter not active, aborted
Traceback (most recent call last):
  File "/Users/grzanka/PycharmProjects/pytripgui/pytripgui/add_vois_vc/add_vois_cont.py", line 29, in <lambda>
    self.view.add_voi_button.emit_on_click(lambda: self._create_add_voi_dialog())
  File "/Users/grzanka/PycharmProjects/pytripgui/pytripgui/add_vois_vc/add_vois_cont.py", line 38, in _create_add_voi_dialog
    controller = AddSingleVOIController(self.model.ctx, view)
  File "/Users/grzanka/PycharmProjects/pytripgui/pytripgui/add_vois_vc/add_single_voi_vc/add_single_voi_cont.py", line 18, in __init__
    self._reload_voi()
  File "/Users/grzanka/PycharmProjects/pytripgui/pytripgui/add_vois_vc/add_single_voi_vc/add_single_voi_cont.py", line 37, in _reload_voi
    self.view.voi_layout.insertWidget(0, voi())
  File "/Users/grzanka/PycharmProjects/pytripgui/pytripgui/add_vois_vc/voi_widget.py", line 52, in __init__
    super().__init__(path)
  File "/Users/grzanka/PycharmProjects/pytripgui/pytripgui/add_vois_vc/voi_widget.py", line 15, in __init__
    uic.loadUi(path, self)
  File "/Users/grzanka/PycharmProjects/pytripgui/venv/lib/python3.8/site-packages/PyQt5/uic/__init__.py", line 238, in loadUi
    return DynamicUILoader(package).loadUi(uifile, baseinstance, resource_suffix)
  File "/Users/grzanka/PycharmProjects/pytripgui/venv/lib/python3.8/site-packages/PyQt5/uic/Loader/loader.py", line 66, in loadUi
    return self.parse(filename, resource_suffix)
  File "/Users/grzanka/PycharmProjects/pytripgui/venv/lib/python3.8/site-packages/PyQt5/uic/uiparser.py", line 1020, in parse
    document = parse(filename)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/xml/etree/ElementTree.py", line 1202, in parse
    tree.parse(source, parser)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/xml/etree/ElementTree.py", line 584, in parse
    source = open(source, "rb")
FileNotFoundError: [Errno 2] No such file or directory: '/Users/grzanka/PycharmProjects/pytripgui/./view/spherical_voi.ui'
Abort trap: 6
```

Reply:
```
(venv) Leszeks-Air:pytripgui grzanka$ ls -al pytripgui/view/spherical_voi.ui 
-rw-r--r--  1 grzanka  staff  10705 Sep 16 14:26 pytripgui/view/spherical_voi.ui
(venv) Leszeks-Air:pytripgui grzanka$ ls -al /Users/grzanka/PycharmProjects/pytripgui/./view/spherical_voi.ui
ls: /Users/grzanka/PycharmProjects/pytripgui/./view/spherical_voi.ui: No such file or directory
(venv) Leszeks-Air:pytripgui grzanka$ ls -al /Users/grzanka/PycharmProjects/pytripgui/pytripgui/view/spherical_voi.ui
-rw-r--r--  1 grzanka  staff  10705 Sep 16 14:26 /Users/grzanka/PycharmProjects/pytripgui/pytripgui/view/spherical_voi.ui
```

Reply:
The files are not in the same dir. Would sys.path[0] be okay? If not, it would probably be a good idea to create some project_root constant in a config file.

Reply:
I'e already  suggested solution here:
https://github.com/pytrip/pytripgui/pull/459#discussion_r710065462

```
current_directory = os.path.dirname(os.path.realpath(__file__))
```

See also:
https://stackoverflow.com/questions/9271464/what-does-the-file-variable-mean-do
https://www.blog.pythonlibrary.org/2013/10/29/python-101-how-to-find-the-path-of-a-running-script/
https://github.com/pyinstaller/pyinstaller/issues/1598

Reply:
The files are not in the same directory, so I'd have to do something like this:
```suggestion
        os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
```
which really doesn't seem like a good idea especially if in the future we'd have even more nested modules.

Reply:
See also https://pyinstaller.readthedocs.io/en/stable/runtime-information.html

Reply:
Move ui and py files together in same directory